snafu-derive/src/lib.rs

```diff
@@ -1339,492 +1422,315 @@ impl<'a> quote::ToTokens for ContextSelectors<'a> {
     }
 }
 
-struct ContextSelector<'a>(&'a EnumInfo, &'a VariantInfo);
+struct ContextSelector<'a>(&'a EnumInfo, &'a FieldContainer);
```

Comment:
I don't see a problem here, but if you'd help me understand it a bit better, I'd appreciate it!

The part of SNAFU I have the most trouble understanding is the code generation, both because I haven't worked in that area of SNAFU and because I haven't done much of it in other crates.  So, I'm sorry if this is pretty basic.

What does this `ContextSelector` do?  I see that it's specific to enums, as is `ContextSelectors`, and I don't see something similar for `ToTokens` for structs.  Is this related to generating the selector structs for the enum variants?  I'm guessing so, because the `generate_snafu` implementation for enums relies heavily on `ContextSelectors`, whereas for structs it does a lot more work directly in the function using the fields and field containers, to build up a `ContextSelector` in the end.

Reply:
> The part of SNAFU I have the most trouble understanding is the code generation, both because I haven't worked in that area of SNAFU and because I haven't done much of it in other crates. 

It's also one of the more gnarly parts of this code, IMO. I always have to spend a bit of time thinking about it.

> What does this `ContextSelector` do? I see that it's specific to enums, as is `ContextSelectors`, and I don't see something similar for `ToTokens` for structs. Is this related to generating the selector structs for the enum variants? I'm guessing so, because the `generate_snafu` implementation for enums relies heavily on `ContextSelectors`, whereas for structs it does a lot more work directly in the function using the fields and field containers, to build up a `ContextSelector` in the end.

The `quote!` macro can be used like `quote! { #something }`. Anything that implements [`ToTokens`](https://docs.rs/quote/1.0.7/quote/trait.ToTokens.html) can be used for interpolation. This `ContextSelector` basically just exists as a place to implement `ToTokens` on.

Previously, I had extracted this code in an attempt to help organize it. However, that organization wasn't quite what was needed to to support structs. I re-extracted it into the types in the `shared` module. The struct code uses those new types directly, and the enum code still goes through the old types. It seems likely that I'll try to unify those two code paths a bit more, but that's mostly refactoring and I didn't want to get stuck in a refactoring hole forever. :-)